# -*- coding: utf-8 -*-
# @Author: Dongqing Sun
# @E-mail: Dongqingsun96@gmail.com
# @Date:   2021-03-27 23:22:54
# @Last Modified by:   Dongqing Sun
# @Last Modified time: 2021-03-30 19:32:34


import os
import argparse as ap
import numpy as np
import scipy.sparse as sp_sparse
import re

from MAESTRO.scATAC_H5Process import *
from MAESTRO.scRNA_utility import RSCRIPT_PATH


def multiomeqc_parser(subparsers):
    """
    Add main function init-scatac argument parsers.
    """

    workflow = subparsers.add_parser("multiome-qc", 
        help = "Perform quality control for scATAC-seq peak-cell binary matrix. "
        "Filter cells according to number of peaks covered.")
    group_input = workflow.add_argument_group("Input arguments")
    group_input.add_argument("--format", dest = "format", default = "", 
        choices = ["h5", "mtx", "plain"], 
        help = "Format of the count matrix file. Please make the format of peak and RNA counts consistent. ")
    group_input.add_argument("--peakcount", dest = "peakcount", default = "", 
        help = "Location of peak count matrix file. "
        "Peak count matrix with peaks as rows and cells as columns. "
        "If the format is 'h5' or 'plain', users need to specify the name of the count matrix file "
        "and row names should be like 'chromosome_peakstart_peakend', such as 'chr10_100020591_100020841'. "
        "If the format is 'mtx', the 'matrix' should be the name of .mtx formatted matrix file, such as 'matrix.mtx'.")
    group_input.add_argument("--genecount", dest = "genecount", default = "", 
        help = "Location of gene count matrix file. "
        "If the format is 'h5' or 'plain', users need to specify the name of the count matrix file. "
        "If the format is 'mtx', the 'matrix' should be the name of .mtx formatted matrix file, such as 'matrix.mtx'.")
    group_input.add_argument("--atac-qc", dest = "atac_qc", default = "", 
        help = "Location of single-cell-level QC file for ATAC generated by MAESTRO pipeline (singlecell.txt).")
    group_input.add_argument("--rna-qc", dest = "rna_qc", default = "", 
        help = "Location of single-cell-level QC file for RNA generated by MAESTRO pipeline (*_count_gene_stat.txt)")
    group_input.add_argument("--separator", dest = "separator", default = "tab", 
        choices = ["tab", "space", "comma"],
        help = "The separating character (only for the format of 'plain'). "
        "Values on each line of the plain matrix file will be separated by the character. DEFAULT: tab.")
    group_input.add_argument("--atac-feature", dest = "atac_feature", default = "", 
        help = "Location of feature file (required for the format of 'mtx'). "
        "Features correspond to row indices of count matrix. "
        "The feature file should be the peak bed file with 3 columns. For example, peaks.bed.")
    group_input.add_argument("--atac-barcode", dest = "atac_barcode", default = "", 
        help = "Location of barcode file (required for the format of 'mtx'). "
        "Cell barcodes correspond to column indices of count matrix. For example, barcodes.tsv. ")
    group_input.add_argument("--rna-feature", dest = "rna_feature", default = "features.tsv", 
        help = "Location of feature file (required for the format of 'mtx'). "
        "Features correspond to row indices of count matrix. DEFAULT: features.tsv.")
    group_input.add_argument("--gene-column", dest = "gene_column", default = 2, type = int,
        help = "If the format is 'mtx', please specify which column of the feature file to use for gene names. DEFAULT: 2.")
    group_input.add_argument("--rna-barcode", dest = "rna_barcode", default = "barcodes.tsv", 
        help = "Location of barcode file (required for the format of 'mtx'). "
        "Cell barcodes correspond to column indices of count matrix. DEFAULT: barcodes.tsv. ")
    group_input.add_argument("--species", dest = "species", default = "GRCh38", 
        choices = ["GRCh38", "GRCm38"], type = str, 
        help = "Species (GRCh38 for human and GRCm38 for mouse). DEFAULT: GRCh38.") 

    # Quality control cutoff
    group_cutoff = workflow.add_argument_group("Quality control arguments")
    group_cutoff.add_argument("--rna-count-cutoff", dest = "rna_count_cutoff", default = 1000, type = int,
        help = "Cutoff for the number of count in each cell for RNA library. DEFAULT: 1000.")
    group_cutoff.add_argument("--rna-gene-cutoff", dest = "rna_gene_cutoff", default = 500, type = int,
        help = "Cutoff for the number of genes included in each cell. DEFAULT: 500.")
    group_cutoff.add_argument("--rna-cell-cutoff", dest = "rna_cell_cutoff", default = 10, type = int,
        help = "Cutoff for the number of cells covered by each gene. DEFAULT: 10.")  
    group_cutoff.add_argument("--atac-peak-cutoff", dest = "atac_peak_cutoff", default = 100, type = int,
        help = "Minimum number of peaks included in each cell. DEFAULT: 100.")
    group_cutoff.add_argument("--atac-count-cutoff", dest = "atac_count_cutoff", default = 1000, type = int, 
        help = "Cutoff for the number of count in each cell for ATAC library. ")
    group_cutoff.add_argument("--atac-frip-cutoff", dest = "atac_frip_cutoff", default = 0.2, type = float, 
        help = "Cutoff for fraction of reads in promoter in each cell. It works only when --atac-qc is provided. DEFAULT: 0.2.")
    group_cutoff.add_argument("--atac-cell-cutoff", dest = "atac_cell_cutoff", default = 10, type = int,
        help = "Minimum number of cells covered by each peak. DEFAULT: 10.")    

    group_output = workflow.add_argument_group("Output arguments")
    group_output.add_argument("-d", "--directory", dest = "directory", default = "MAESTRO", 
        help = "Path to the directory where the result file shall be stored. DEFAULT: MAESTRO.")
    group_output.add_argument("--outprefix", dest = "outprefix", default = "10x-genomics", 
        help = "Prefix of output files. DEFAULT: MAESTRO.")


def multiome_qc(fileformat, directory, outprefix, peakcount, genecount, atac_qc, rna_qc, atac_feature, atac_barcode, 
    rna_feature, gene_column, rna_barcode, species):
    try:
        os.makedirs(directory)
    except OSError:
        # either directory exists (then we can ignore) or it will fail in the
        # next step.
        pass

    if fileformat == "plain":
        matrix_dict = read_count(peakcount, separator)
        peakmatrix = matrix_dict["matrix"]
        peakmatrix = sp_sparse.csc_matrix(peakmatrix, dtype=np.int32)
        atac_features = matrix_dict["features"]
        atac_features = [f.encode() for f in atac_features]
        atac_barcodes = matrix_dict["barcodes"]

        matrix_dict = read_count(genecount, separator)
        genematrix = matrix_dict["matrix"]
        genematrix = sp_sparse.csc_matrix(genematrix, dtype=numpy.float32)
        rna_features = matrix_dict["features"]
        rna_barcodes = matrix_dict["barcodes"]

    elif fileformat == "h5":
        scatac_count = read_10X_h5(peakcount)
        peakmatrix = scatac_count.matrix
        atac_features = scatac_count.names.tolist()
        atac_features = [re.sub("\W", "_", feature.decode()) for feature in atac_features]
        atac_features = [feature.encode() for feature in atac_features]
        atac_barcodes = scatac_count.barcodes.tolist()

        if type(atac_features[0]) == bytes:
            atac_features = [i.decode() for i in atac_features]
        if type(atac_barcodes[0]) == bytes:
            atac_barcodes = [i.decode() for i in atac_barcodes]

        scrna_count = read_10X_h5(genecount)
        genematrix = scrna_count.matrix
        rna_features = scrna_count.names.tolist()
        rna_barcodes = scrna_count.barcodes.tolist()

        if type(rna_features[0]) == bytes:
            rna_features = [i.decode() for i in rna_features]
        if type(rna_barcodes[0]) == bytes:
            rna_barcodes = [i.decode() for i in rna_barcodes]

    elif fileformat == "mtx":
        matrix_dict = read_10X_mtx(matrix_file = peakcount, feature_file = atac_feature, barcode_file = atac_barcode, datatype = "Peak")
        peakmatrix = matrix_dict["matrix"]
        atac_features = matrix_dict["features"]
        atac_features = [f.encode() for f in atac_features]
        atac_barcodes = matrix_dict["barcodes"]

        matrix_dict = read_10X_mtx(matrix_file = genecount, feature_file = rna_feature, barcode_file = rna_barcode, datatype = "Gene", gene_column = gene_column)
        genematrix = matrix_dict["matrix"]
        rna_features = matrix_dict["features"]
        rna_barcodes = matrix_dict["barcodes"]

    barcode_overlapped = list(set(rna_barcodes) & set(atac_barcodes))
    rna_barcode_idx = [rna_barcodes.index(i) for i in barcode_overlapped]
    atac_barcode_idx = [atac_barcodes.index(i) for i in barcode_overlapped]

    genematrix_filtered = genematrix[:, np.array(rna_barcode_idx)]
    peakmatrix_filtered = peakmatrix[:, np.array(atac_barcode_idx)]

    all_feature_matrix = sp_sparse.vstack([genematrix_filtered, peakmatrix_filtered])
    all_features = rna_features + atac_features
    feature_types = ["Gene Expression"]*len(rna_features) + ["Peaks"]*len(atac_features)

    write_10X_h5(os.path.join(directory, outprefix + "_joint_filtered_gene_count.h5"), matrix = genematrix_filtered, features = rna_features, barcodes = barcode_overlapped, genome = species, datatype = 'Gene')
    write_10X_h5(os.path.join(directory, outprefix + "_joint_filtered_peak_count.h5"), matrix = peakmatrix_filtered, features = atac_features, barcodes = barcode_overlapped, genome = species, datatype = 'Peak')
    write_10X_h5(filename = os.path.join(directory, outprefix + "_joint_filtered_multiome_feature_count.h5"), matrix = all_feature_matrix, features = all_features, barcodes = barcode_overlapped, feature_types = feature_types, genome = species, datatype = 'Multiome')

    multiome_count_file = os.path.join(".", outprefix + "_joint_filtered_multiome_feature_count.h5")
    cmd = "Rscript %s/Multiome_qc_filtering.R --prefix %s --outdir %s --rna_statpath %s --atac_statpath  %s --count %s" %(RSCRIPT_PATH, outprefix, directory, rna_qc, atac_qc, multiome_count_file)
    os.system(cmd)
