<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Gali Bai" />

<meta name="date" content="2021-07-19" />

<title>snRNA-seq</title>

<script src="site_libs/header-attrs-2.9/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MAESTRO 1.5.1</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    scRNA-seq
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="scrna.html">10X PBMC 1k</a>
    </li>
    <li>
      <a href="scrna_stepbystep.html">10X PBMC 1k Custom Analysis</a>
    </li>
    <li>
      <a href="snrna.html">10X snRNA-seq</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    scATAC-seq
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="scatac.html">10X PBMC 500</a>
    </li>
    <li>
      <a href="scatac_stepbystep.html">10X PBMC 500 Custom Analysis</a>
    </li>
    <li>
      <a href="sci_atac.html">sci-ATAC-seq</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    multi-sample scRNA-seq
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="multi-scrna.html">10X multi-sample</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    multi-sample scATAC-seq
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="multi-scatac.html">time-dependent multi-sample</a>
    </li>
    <li>
      <a href="multi-scatac_stepbystep.html">Custom Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a>
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/liulab-dfci/MAESTRO">Source code</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">snRNA-seq</h1>
<h4 class="author">Gali Bai</h4>
<h4 class="date">2021-07-19</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-07-27
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>MAESTRO_documentation/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201223code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201223)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201223code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201223)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcombaigal628MultiSamplescATACseqtree4a7b06feb4cf21a05554efbc3bf8fc9c17b37683targetblank4a7b06fa"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/baigal628/MultiSample_scATACseq/tree/4a7b06feb4cf21a05554efbc3bf8fc9c17b37683" target="_blank">4a7b06f</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcombaigal628MultiSamplescATACseqtree4a7b06feb4cf21a05554efbc3bf8fc9c17b37683targetblank4a7b06fa" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/baigal628/MultiSample_scATACseq/tree/4a7b06feb4cf21a05554efbc3bf8fc9c17b37683" target="_blank">4a7b06f</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .RData
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    code/.DS_Store
    Ignored:    data/.DS_Store
    Ignored:    data/multi-scatac/.DS_Store
    Ignored:    data/multi-scrna/

Untracked files:
    Untracked:  .RDataTmp
    Untracked:  code/snRNA_genelength.R
    Untracked:  code/snRNA_genelength_TPM.R
    Untracked:  data/allen/
    Untracked:  data/multi-scatac/bed/
    Untracked:  data/multi-scatac/bigwig/GDF7_bed_added.png
    Untracked:  output/snRNA_TPM_VS_genelength.png
    Untracked:  output/snRNA_genelengthVSLognormUMI.png
    Untracked:  output/snRNA_genelengthVSUMI.png

Unstaged changes:
    Deleted:    MultiSample_scATACseq.Rproj
    Deleted:    atac_pbmc_500_nextgem.GIGGLE/0.peaks.bed
    Deleted:    atac_pbmc_500_nextgem.GIGGLE/0.peaks.bed.gz
    Deleted:    atac_pbmc_500_nextgem.GIGGLE/0.peaks.bed.result.xls
    Modified:   code/Basic_Operations.R
    Modified:   pbmc_1k_v3_8k_res.rds
    Modified:   pbmc_1k_v3_Monocyte_filtered.pdf
    Modified:   pbmc_1k_v3_Monocyte_top.pdf

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>In this example, we will perform a single-nuclei RNA-seq using 10X dataset of 750 Sorted Nuclei from Human Brain. The raw dataset can be downloaded from the 10X Genomics website. We will show you how to run through the whole MAESTRO pipeline from the raw sequencing fastq files to the final results.</p>
</div>
<div id="run-maestro-pipeline" class="section level2">
<h2>Run MAESTRO pipeline</h2>
<div id="step-0.-download-the-data-and-prepare-the-environment" class="section level3">
<h3>Step 0. Download the data and prepare the environment</h3>
<p>Please download the raw data from 10X genomics website.</p>
<pre class="bash"><code>$ wget https://cf.10xgenomics.com/samples/cell-exp/6.0.0/Brain_3p_LT/Brain_3p_LT_fastqs.tar
$ tar xvf Brain_3p_LT_fastqs.tar</code></pre>
<p>Before running MAESTRO, users need to activate the MAESTRO environment.</p>
<pre class="bash"><code>$ conda activate MAESTRO</code></pre>
</div>
<div id="step-1.-configure-the-maestro-workflow" class="section level3">
<h3>Step 1. Configure the MAESTRO workflow</h3>
<p>Initialize the MAESTRO scRNA-seq workflow using MAESTRO scrna-init command. This will install a Snakefile and a config file in this directory.</p>
<pre class="bash"><code>MAESTRO scrna-init \
--platform 10x-genomics --species GRCh38 \
--STARsolo_Features GeneFull \
--STARsolo_threads 12 \
--cores 32 --rseqc --directory Brain_3p_LT \
--count-cutoff 1000 --gene-cutoff 500 --cell-cutoff 10 \
--mapindex references/Refdata_scRNA_MAESTRO_GRCh38_1.2.2/GRCh38_STAR_2.7.6a \
--whitelist references/whitelist/3M-february-2018.txt \
--barcode-start 1 --barcode-length 16 --umi-start 17 --umi-length 12 \
--lisadir references/hg38_1000_2.0.h5 \
--signature human.immune.CIBERSORT</code></pre>
<p>To get a full description of command-line options, please use the command</p>
<pre class="bash"><code>$ MAESTRO scrna-init -h</code></pre>
<div id="here-we-list-all-the-arguments-and-their-description." class="section level4">
<h4>Here we list all the arguments and their description.</h4>
<p><strong>Input files arguments:</strong></p>
<table>
<colgroup>
<col width="45%" />
<col width="55%" />
</colgroup>
<thead>
<tr class="header">
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>--platform</code></td>
<td>{10x-genomics,Dropseq,Smartseq2} Platform of single cell RNA-seq. DEFAULT: 10x-genomics.</td>
</tr>
<tr class="even">
<td><code>--sample-file</code></td>
<td>JSON file with sample file information.</td>
</tr>
<tr class="odd">
<td><code>--fastq-barcode</code></td>
<td>Specify the barcode fastq file, only for the platform of ‘Dropseq’. If there are multiple pairs of fastq, please provide a comma-separated list of barcode fastq files. For example, <code>--fastq-barcode test1_1.fastq,test2_1.fastq</code>.</td>
</tr>
<tr class="even">
<td><code>--fastq-transcript</code></td>
<td>Specify the transcript fastq file, only for the platform of ‘Dropseq’.</td>
</tr>
<tr class="odd">
<td><code>--species</code></td>
<td>{GRCh38,GRCm38} Specify the genome assembly (GRCh38 for human and GRCm38 for mouse). DEFAULT: GRCh38.</td>
</tr>
</tbody>
</table>
<p><strong>STARsolo reads mapping arguments:</strong></p>
<table>
<colgroup>
<col width="45%" />
<col width="55%" />
</colgroup>
<thead>
<tr class="header">
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>--STARsolo_Features</code></td>
<td>{Gene,GeneFull,Gene GeneFull,SJ,Velocyto} Parameters passed to STARsolo –soloFeatures. Specify –soloFeatures Gene for single-cell data. Specify –soloFeatures GeneFull for single-nuclei data. Specify –soloFeatures Gene GeneFull for getting both counts in exons level and exon + intron level (velocity). If multiple features are listed, e.g. Gene GeneFull SJ, only the count matrix generated by the first feature (here: Gene) will be used for downstream analysis. DEFAULT: Gene.</td>
</tr>
<tr class="even">
<td><code>--STARsolo_Threads</code></td>
<td>Threads for running STARsolo. DEFAULT: 12.</td>
</tr>
</tbody>
</table>
<p><strong>Running and output arguments:</strong></p>
<table>
<colgroup>
<col width="45%" />
<col width="55%" />
</colgroup>
<thead>
<tr class="header">
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>--cores</code></td>
<td>The number of cores to use. DEFAULT: 10.</td>
</tr>
<tr class="even">
<td><code>--rseqc</code></td>
<td>Whether or not to run RSeQC. If set, the pipeline will include the RSeQC part and then takes a longer time. By default (not set), the pipeline will skip the RSeQC part.</td>
</tr>
<tr class="odd">
<td><code>--directory</code></td>
<td>Path to the directory where the workflow shall be initialized and results shall be stored. DEFAULT: MAESTRO.</td>
</tr>
<tr class="even">
<td><code>--mergedname</code></td>
<td>Prefix of merged output files. DEFAULT: All_sample.</td>
</tr>
<tr class="odd">
<td><code>--outprefix</code></td>
<td>Prefix of output files. Only required for platform of ‘Smartseq2’. DEFAULT: MAESTRO.</td>
</tr>
</tbody>
</table>
<p><strong>Quality control arguments:</strong></p>
<table>
<colgroup>
<col width="45%" />
<col width="55%" />
</colgroup>
<thead>
<tr class="header">
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>--count-cutoff</code></td>
<td>Cutoff for the number of count in each cell. DEFAULT: 1000.</td>
</tr>
<tr class="even">
<td><code>--gene-cutoff</code></td>
<td>Cutoff for the number of genes included in each cell. DEFAULT: 500.</td>
</tr>
<tr class="odd">
<td><code>--cell-cutoff</code></td>
<td>Cutoff for the number of cells covered by each gene. DEFAULT: 10.</td>
</tr>
</tbody>
</table>
<p><strong>Reference genome arguments:</strong></p>
<table>
<colgroup>
<col width="45%" />
<col width="55%" />
</colgroup>
<thead>
<tr class="header">
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>--mapindex</code></td>
<td>Genome index directory for STAR. Users can just download the index file for <a href="http://cistrome.org/~galib/MAESTRO/references/scRNA/Refdata_scRNA_MAESTRO_GRCh38_1.2.2.tar.gz">human</a> and <a href="http://cistrome.org/~galib/MAESTRO/references/scRNA/Refdata_scRNA_MAESTRO_GRCm38_1.2.2.tar.gz">mouse</a> from CistromeDB and decompress them. Then specify the index directory for STAR, for example, <code>--mapindex Refdata_scRNA_MAESTRO_GRCh38_1.1.0/GRCh38_STAR_2.7.3a</code>.</td>
</tr>
<tr class="even">
<td><code>--rsem</code></td>
<td>The prefix of transcript references for RSEM used by rsem-prepare-reference (Only required when the platform is Smartseq2). Users can directly download the reference file for <a href="http://cistrome.org/~galib/MAESTRO/references/scRNA/Refdata_scRNA_MAESTRO_GRCh38_1.1.0.tar.gz">human</a> and <a href="http://cistrome.org/~galib/MAESTRO/references/scRNA/Refdata_scRNA_MAESTRO_GRCm38_1.1.0.tar.gz">mouse</a> from CistromeDB and decompress them. Then specify the prefix for RSEM, for example, <code>--rsem Refdata_scRNA_MAESTRO_GRCh38_1.1.0/GRCh38_RSEM_1.3.2/GRCh38</code>.</td>
</tr>
</tbody>
</table>
<p><strong>Barcode arguments, for platform of ‘Dropseq’ or ‘10x-genomics’:</strong></p>
<table>
<colgroup>
<col width="45%" />
<col width="55%" />
</colgroup>
<thead>
<tr class="header">
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>--whitelist</code></td>
<td>If the platform is ‘Dropseq’ or ‘10x-genomics’, please specify the barcode library (whitelist) so that STARsolo can do the error correction and demultiplex of cell barcodes. The 10X Chromium whitelist file can be found inside the CellRanger distribution. Please make sure that the whitelist is compatible with the specific version of the 10X chemistry: V2 or V3. For example, in CellRanger 3.1.0, the V2 whitelist is ‘cellranger-3.1.0/cellranger-cs/3.1.0/lib/python/cellranger/barcodes/737K-august-2016.txt’. The V3 whitelist is ‘cellranger-3.1.0/cellranger-cs/3.1.0/lib/python/cellranger/barcodes/3M-february-2018.txt’.</td>
</tr>
<tr class="even">
<td><code>--barcode-start</code></td>
<td>The start site of each barcode. DEFAULT: 1.</td>
</tr>
<tr class="odd">
<td><code>--barcode-length</code></td>
<td>The length of cell barcode. For 10x-genomics, the length of barcode is 16. DEFAULT: 16.</td>
</tr>
<tr class="even">
<td><code>--umi-start</code></td>
<td>The start site of UMI. DEFAULT: 17.</td>
</tr>
<tr class="odd">
<td><code>--umi-length</code></td>
<td>The length of UMI. For 10x-genomics, the length of V2 chemistry is 10. For 10X V3 chemistry, the length is 12. DEFAULT: 10.</td>
</tr>
<tr class="even">
<td><code>--trimR1</code></td>
<td>Whether or not to run the R1 file. If set, the pipeline will include the trim off anything after the R1 reads past barcode information. Only necessary if reads were sequenced past these barcodes, by default not set.</td>
</tr>
</tbody>
</table>
<p><strong>Regulator identification arguments:</strong></p>
<table>
<colgroup>
<col width="45%" />
<col width="55%" />
</colgroup>
<thead>
<tr class="header">
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>--lisadir</code></td>
<td>Path to the LISA data files. Please download LISA’s required data from cistrome.org: <a href="http://cistrome.org/~alynch/data/lisa_data/hg38_1000_2.0.h5">human</a> and <a href="http://cistrome.org/~alynch/data/lisa_data/mm10_1000_2.0.h5">mouse</a>. The latest version of <a href="https://github.com/liulab-dfci/lisa2">LISA2</a> was recently released to largely decrease processing time on multiple gene lists. Please check out for detailed information.</td>
</tr>
</tbody>
</table>
<p><strong>Cell signature arguments:</strong></p>
<table>
<colgroup>
<col width="45%" />
<col width="55%" />
</colgroup>
<thead>
<tr class="header">
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>--signature</code></td>
<td>Cell signature file used to annotate cell types. MAESTRO provides several sets of built-in cell signatures. Users can choose from [‘human.immune.CIBERSORT’, ‘mouse.brain.ALLEN’, ‘mouse.all.facs.TabulaMuris’, ‘mouse.all.droplet.TabulaMuris’]. Custom cell signatures are also supported. In this situation, users need to provide the file location of cell signatures, and the signature file is tab-separated without header. The first column is cell type, and the second column is signature gene. DEFAULT: human.immune.CIBERSORT.</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="step-2.-configure-the-samples.json-file" class="section level3">
<h3>Step 2. Configure the samples.json file</h3>
<p>In the project directory, we will next run <code>MAESTRO samples-init</code> subcommand to configure samples.json file. The purpose of this step is let MAESTRO read the right format and path of input data.</p>
<pre class="bash"><code>$ cd Brain_3p_LT/

$ MAESTRO samples-init --assay_type scrna --platform 10x-genomics --data_typ fastq --data_dir MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs</code></pre>
<p>To get a full description of command-line options, please use the command</p>
<pre class="bash"><code>MAESTRO samples-init -h</code></pre>
<p><strong>Input files arguments:</strong></p>
<table>
<colgroup>
<col width="45%" />
<col width="55%" />
</colgroup>
<thead>
<tr class="header">
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>--assay_type</code></td>
<td>{scrna,scatac} choose the type of assays.</td>
</tr>
<tr class="even">
<td><code>--platform</code></td>
<td>{10x-genomics,sci-ATAC-seq,microfluidic} choose the type of sequencing platform.</td>
</tr>
<tr class="odd">
<td><code>--data_type</code></td>
<td>{fastq,fragment,bam} choose the type of input data. When using fastq format as input, MAESTRO can only recognize the 10X-like naming: [Sample Name]<em>S1_L00[Lane Number]</em>[Read Type]_001.fastq.gz.</td>
</tr>
<tr class="even">
<td><code>--data_dir</code></td>
<td>FULL path to the input file folder.</td>
</tr>
</tbody>
</table>
<p>Once configured, you will find a samples.json file in the current project directory:</p>
<pre class="bash"><code>$ cat samples.json
{
    &quot;mm_LT_750&quot;: {
        &quot;I1&quot;: [
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L001_I1_001.fastq.gz&quot;,
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L002_I1_001.fastq.gz&quot;,
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L003_I1_001.fastq.gz&quot;,
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L004_I1_001.fastq.gz&quot;
        ],
        &quot;I2&quot;: [
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L001_I2_001.fastq.gz&quot;,
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L002_I2_001.fastq.gz&quot;,
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L003_I2_001.fastq.gz&quot;,
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L004_I2_001.fastq.gz&quot;
        ],
        &quot;R1&quot;: [
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S15_L001_R1_001.fastq.gz&quot;,
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L002_R1_001.fastq.gz&quot;,
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L003_R1_001.fastq.gz&quot;,
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L004_R1_001.fastq.gz&quot;
        ],
        &quot;R2&quot;: [
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L001_R2_001.fastq.gz&quot;,
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L002_R2_001.fastq.gz&quot;,
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L003_R2_001.fastq.gz&quot;,
            &quot;/liulab/galib/sc_CIDC/MAESTRO/test/snRNA/data/Brain_3p_LT_fastqs/mm_LT_750_S5_L004_R2_001.fastq.gz&quot;
        ]
    }
}
</code></pre>
</div>
<div id="step-3.-run-snakemake-pipeline" class="section level3">
<h3>Step 3. Run snakemake pipeline</h3>
<p>Before running the workflow, please check the <code>config.yaml</code> and see if it is configured correctly. Once <code>config.yaml</code> is configured, users can use snakemake to run the workflow.</p>
<pre class="bash"><code>$ snakemake -np

$ nohup snakemake -j 16 &gt; run.out &amp;
##Instead of using nohup, submitting $ snakemake -j 16 by slurm is highly recommended.</code></pre>
</div>
<div id="step-4.-understand-the-final-output-files" class="section level3">
<h3>Step 4. Understand the final output files</h3>
<p>The whole pipeline in this example takes about 3 hours with 12 cores. If MAESTRO runs successfully, an output folder will contain several useful outputs as described below.</p>
<pre class="bash"><code>$ ls Result
Analysis  Benchmark  Log  QC  Report  STAR</code></pre>
<div id="output-files" class="section level4">
<h4>Output files</h4>
<ul>
<li><strong>STAR/:</strong> The <code>STAR</code> directory contains all the mapping and analysis files generated by <code>STAR</code> normal (Smartseq2) or solo mode (10x-genomics or Dropseq). For solo mode, <code>Solo.out/Gene/</code> stores raw and filtered count matrix. In MAESTRO pipeline, raw count matrix is used for a further filter according to quality control arguments like <code>--count-cutoff</code> <code>--gene-cutoff</code> and <code>--cell-cutoff</code>.</li>
<li><strong>QC/:</strong> The <code>QC</code> directory contains quality control analysis results of scRNA-seq data, including the filtered count matrix <code>outprefix_filtered_gene_count.h5</code> and RSeQC results (if <code>--rseqc</code> is set).</li>
<li><strong>Analysis/:</strong> The <code>Analysis</code> directory contains a Seurat R object, as well as clustering result, cell type annotation result and driver transcription factor identification result, which we will introduce in the step-by-step analysis.</li>
<li><strong>Benchmark/:</strong> The <code>Benchmark</code> directory stores benchmark files for all rules in Snakefile, each of which contains a tab-separated table of run times and memory usage in MiB.</li>
<li><strong>Log/:</strong> The Log directory contains the log files generated in the pipeline analysis.</li>
<li><strong>Report/:</strong> Contain the <code>{outprefix}_scRNA_report.html</code> file summarizes all the results in an HTML based document. The summary HTML for the example can be found <a href="http://cistrome.org/~galib/MAESTRO/MAESTRO_report/pbmc_1k_v3_scRNA_report.html">here</a>.</li>
</ul>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.0 (2021-05-18)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6        knitr_1.33        magrittr_2.0.1    R6_2.5.0         
 [5] rlang_0.4.11      fansi_0.5.0       stringr_1.4.0     tools_4.1.0      
 [9] xfun_0.24         utf8_1.2.1        git2r_0.28.0      jquerylib_0.1.4  
[13] htmltools_0.5.1.1 ellipsis_0.3.2    rprojroot_2.0.2   yaml_2.2.1       
[17] digest_0.6.27     tibble_3.1.2      lifecycle_1.0.0   crayon_1.4.1     
[21] later_1.2.0       sass_0.4.0        vctrs_0.3.8       promises_1.2.0.1 
[25] fs_1.5.0          glue_1.4.2        evaluate_0.14     rmarkdown_2.9    
[29] stringi_1.6.2     bslib_0.2.5.1     compiler_4.1.0    pillar_1.6.1     
[33] jsonlite_1.7.2    httpuv_1.6.1      pkgconfig_2.0.3  </code></pre>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
