#' Read distribution plot for scATAC-seq analysis
#'
#' Summary plot of total reads, duplicate reads ratio, mappability, mitochondria and peak reads ratio for scATAC-seq analysis.
#'
#' @docType methods
#' @name ATACReadDistrPlot
#' @rdname ATACReadDistrPlot
#'
#' @param stat.filepath Path of the single-cell statistic file generated by \code{MAESTRO}.
#' @param name Name for the output read distribution plot.
#' @param mapping Alingment tool used for scATAC-seq 
#' @author Chenfei Wang, Dongqing Sun
#'
#' @importFrom ggplot2 aes element_blank element_text geom_bar geom_text ggplot margin theme theme_set xlab ylab
#' @export

ATACReadDistrPlot <- function(stat.filepath, mapping, name)
{
  theme_set(cowplot::theme_cowplot())
  RCB_blue = "#2166AC"
  
if(mapping == "minimap2"){
    # singlecell = read.csv(stat.filepath)
    # singlecell_fragsum = colSums(singlecell[,c(2:8,10:18)])
    # singlecell_readsum = singlecell_fragsum*2
    # map_list = c(singlecell_readsum[c(1:2,7,6,15,12)])/1000000
    bam_stat_list = readLines(stat.filepath)[c(1,4,14,15,17,16)]
    total_reads = unlist(strsplit(bam_stat_list[1], split = " "))[1]
    dup_reads = unlist(strsplit(bam_stat_list[2], split = " "))[1]
    map_list = as.integer(c(total_reads, dup_reads, bam_stat_list[3:6]))/1000000
    
    map_ratio_list = round(map_list*100/map_list[1],2)
    map_ratio_list = paste0(map_ratio_list, "%")
    Group = c("Total reads", "Duplicate reads", "Uniquely mapped reads","Mitochondria reads", "Peak reads", "Promoter reads")
    map_df = data.frame(ReadCount = map_list, Group, stringsAsFactors = FALSE)
    map_df$Group = factor(map_df$Group, levels = Group)
    
    png(paste0(name ,"_scATAC_read_distr.png"), width=4.5,height=4.8, res = 300, units = "in")
    p = ggplot(map_df, aes(x = Group, y = ReadCount)) + 
      geom_bar(stat="identity", fill = RCB_blue) + 
      xlab(NULL) + ylab("Read counts (M)")  + 
      theme(axis.title = element_text(size = 13), axis.text.y = element_text(size = 12, angle = 90, hjust = 0.5), 
            axis.text.x = element_text(angle = 45, size = 12, hjust = 1), axis.ticks.x = element_blank(),
            legend.position="none", axis.title.y = element_text(margin = margin(r = 20, unit = "pt")))+ 
      geom_text(aes(label = map_ratio_list, y = ReadCount + 0.04*map_list[1] ), size = 4.8)
    print(p)
    dev.off()
    }else{
    bam_stat_list = readLines(stat.filepath)[c(1,2,3,4,5)]
    map_list = as.integer(bam_stat_list)/1000000
    
    map_ratio_list = round(map_list*100/map_list[1],2)
    map_ratio_list = paste0(map_ratio_list, "%")
    Group = c("Total reads", "Uniquely mapped reads","Mitochondria reads", "Peak reads", "Promoter reads")
    map_df = data.frame(ReadCount = map_list, Group, stringsAsFactors = FALSE)
    map_df$Group = factor(map_df$Group, levels = Group)
    
    png(paste0(name ,"_scATAC_read_distr.png"), width=4.5,height=4.8, res = 300, units = "in")
    p = ggplot(map_df, aes(x = Group, y = ReadCount)) + 
      geom_bar(stat="identity", fill = RCB_blue) + 
      xlab(NULL) + ylab("Read counts (M)")  + 
      theme(axis.title = element_text(size = 13), axis.text.y = element_text(size = 12, angle = 90, hjust = 0.5), 
            axis.text.x = element_text(angle = 45, size = 12, hjust = 1), axis.ticks.x = element_blank(),
            legend.position="none", axis.title.y = element_text(margin = margin(r = 20, unit = "pt")))+ 
      geom_text(aes(label = map_ratio_list, y = ReadCount + 0.04*map_list[1] ), size = 4.8)
    print(p)
    dev.off()
    }
}